### ISSUE TITLE
softmax sliently generate a wrong output when the mask has an  incompatible shape

### ISSUE THREAD

**Issue by AGFACBNNR** [keras-team/keras#19722]
As shown in this [colab](https://colab.research.google.com/drive/1B5c-9QEwn8GWvi0kOp_mPvSt5BGV34tM?usp=sharing), the mask can arbitrarily change the shape of output.
Although the document requires that the mask has the same shape with input, no check exists  actually.
In my project, the shape of input is hard to compute and this potential incompatibility can affect the result and is hard to debug since softmax now silently generates an wrong output.
Currently the output is:
![1715915144162](https://github.com/keras-team/keras/assets/20224019/53fc812d-64a6-4054-a4f7-f677c7943d42)
You can identify the shape changed by mask(line 2 and line 3 of output), which is different from the correct shape computed by compute_output_shape() (line4).

**Comment by SuryanarayanaY** [comment]
Hi @AGFACBNNR ,

Thanks for reporting this issue. I acknowledge the missing check for same shape for input and mask. Attached [gist](https://colab.sandbox.google.com/gist/SuryanarayanaY/8e41a4f7e80c6707731b5fde975e784e/19722_softmax_maskshape.ipynb#scrollTo=-sr9CgG8jsns) for same.

Will propose a fix for same. Thanks!

**Comment by hertschuh** [comment]
For the record, the bug is that the broadcast happens both ways, if mask is "bigger" than the inputs, the inputs are broadcast. This is weird and we should disallow that.

But the mask doesn't need to be the same shape as the inputs, it just needs to be broadcastable to the input shape. The documentation in the docstring about this is incorrect.
